File: N_M.java
import java.util.concurrent.*;

public class N_M {
    public static void main(String[] args) throws Exception {

        int n = 1000; //escolhe a quantidade
        int tempoTrabalhoMs = 10;
        int tamanhoPool = Runtime.getRuntime().availableProcessors(); //threads do meu pc
        System.out.println("modelo N:M");
        System.out.println("threads " + n);
        System.out.println("threads sistema: " + tamanhoPool);
        ExecutorService pool = Executors.newFixedThreadPool(tamanhoPool);
        long inicio = System.nanoTime(); //comeco
        // manda as n tarefas para o pool fixo
        for (int i = 0; i < n; i++) {
            pool.submit(() -> {
                try { Thread.sleep(tempoTrabalhoMs); }
                catch (InterruptedException e) { Thread.currentThread().interrupt(); }
            });
        }
        pool.shutdown();
        pool.awaitTermination(1, TimeUnit.MINUTES);
        long fim = System.nanoTime(); //fim
        long tempoTotalMs = (fim - inicio) / 1_000_000;

        System.out.println("Tempo total: " + tempoTotalMs + " ms");
    }
}

File: N_M.py
import time
import concurrent.futures

def main():
    n = 1000 #tem que escolher aqui tb
    tempo_trabalho = 0.01 
    tamanho_pool = 16 #sao quantas threads eu tenho no meu pc mesmo

    print("modelo N:M ")
    print("threads:", n)
    print("threads sistema", tamanho_pool)

    inicio = time.time()#comeco
    
    #cria um pool fixo de threads reais
    with concurrent.futures.ThreadPoolExecutor(max_workers=tamanho_pool) as executor:
        for _ in range(n):
            executor.submit(time.sleep, tempo_trabalho)

    fim = time.time()  #fim
    tempo_total_ms = (fim - inicio) * 1000 #ms
    print("tempo total: ", tempo_total_ms,"ms")

if __name__ == "__main__":
    main()

